## Import wooldridge library

In [1]:
import wooldridge as woo

## Load dataset

In [2]:
df = woo.data('meap01')
df.head()

dcode  bcode      math4      read4      lunch  enroll     expend  \
0  1010.0   4937  83.300003  77.800003  40.599998     468  2747475.0   
1  2070.0    597  90.300003  82.300003  27.100000     679  1505772.0   
2  2080.0   4860  61.900002  71.400002  41.750000     400  2121871.0   
3  3010.0    790  85.699997  60.000000  12.750000     251  1211034.0   
4  3010.0   1403  77.300003  59.099998  17.080000     439  1913501.0   

         exppp   lenroll    lexpend    lexppp  
0  5870.672852  6.148468  14.826193  8.677725  
1  2217.631836  6.520621  14.224816  7.704195  
2  5304.677734  5.991465  14.567809  8.576344  
3  4824.836426  5.525453  14.006985  8.481532  
4  4358.771973  6.084499  14.464445  8.379946

## (i) Find the largest and smallest values of $math4$. Does this range make sense? Explain.

In [3]:
# Find the smallest and largest values of math4
min_math4 = df['math4'].min()
max_math4 = df['math4'].max()

# Print results
print(f"Smallest value of math4: {min_math4}")
print(f"Largest value of math4: {max_math4}")

Smallest value of math4: 0.0
Largest value of math4: 100.0


Smallest value: typically should be ≥ 0 (or the minimum possible score on the test).

Largest value: should not exceed the maximum score possible (often 100 or the maximum scale of the test).

If the minimum or maximum is outside the expected range, it could indicate data errors, missing codes, or special values?

## (ii) How many schools have a perfect pass rate on the math test? What percentage of the total sample does this represent?


In [4]:
# Count how many schools have a perfect pass rate (math4 == 100)
perfect_pass_count = (df['math4'] == 100).sum()

# Total number of schools in the dataset
total_schools = df.shape[0]

# Percentage of schools with perfect pass rate
perfect_pass_pct = (perfect_pass_count / total_schools) * 100

# Print results
print(f"Number of schools with perfect pass rate: {perfect_pass_count}")
print(f"Percentage of total sample: {perfect_pass_pct:.2f}%")

Number of schools with perfect pass rate: 38
Percentage of total sample: 2.08%


## (iii) How many schools have math pass rates of exactly 50%?

In [5]:
# Count how many schools have exactly 50% pass rate in math4
count_50 = (df['math4'] == 50).sum()

print(f"Number of schools with math4 = 50%: {count_50}")

Number of schools with math4 = 50%: 17


## (iv) Compare the average passing rates for math and reading. Which test is more difficult to pass?

In [6]:
# Calculate the mean pass rates for math and reading
mean_math = df['math4'].mean()
mean_read = df['read4'].mean()

print(f"Average math pass rate: {mean_math:.2f}%")
print(f"Average reading pass rate: {mean_read:.2f}%")

# Compare which subject had lower pass rate
if mean_math < mean_read:
    print("Math test appears to be more difficult (lower pass rate).")
elif mean_math > mean_read:
    print("Reading test appears to be more difficult (lower pass rate).")
else:
    print("Both tests have the same average pass rate.")

Average math pass rate: 71.91%
Average reading pass rate: 60.06%
Reading test appears to be more difficult (lower pass rate).


## (v) Find the correlation between $math4$ and $read4$. What do you conclude?

In [7]:
# Calculate the correlation between math4 and read4
corr_math_read = df['math4'].corr(df['read4'])

print(f"Correlation between math4 and read4: {corr_math_read:.3f}")

Correlation between math4 and read4: 0.843


the correlation is high and positive (0.843), which suggests that schools that do well in reading also do well in math.

## (vi) The variable $exppp$ is spending per student. Find the mean $exppp$ and its standard deviation. Would you say there is a wide variation in spending per student?

In [8]:
# Calculate mean and standard deviation of exppp
mean_exppp = df['exppp'].mean()
std_exppp = df['exppp'].std()

print(f"Average expenditure per pupil: ${mean_exppp:,.2f}")
print(f"Standard deviation: ${std_exppp:,.2f}")

Average expenditure per pupil: $5,194.87
Standard deviation: $1,091.89


The standard deviation is about 21% of the mean (1091 ÷ 5195 ≈ 0.21).

This indicates that there is considerable variation in per-student spending across schools.

### (vii) Suppose school A spends US 6,000 per student and school B spends US 5,500 per student. By what percentage does school A exceed school B's spending? Compare this to 100 • [log(6,000) - log(5,500)], which is the approximate percentage difference based on the difference of the natural logs.

In [9]:
#import numpy
import numpy as np

# Expenditures per pupil
exppp_A = 6000
exppp_B = 5500

# (1) Exact percentage difference
percent_diff = ((exppp_A - exppp_B) / exppp_B) * 100

# (2) Approximate percentage difference using natural logs
log_diff = 100 * (np.log(exppp_A) - np.log(exppp_B))

# Print results
print(f"Exact percentage difference: {percent_diff:.2f}%")
print(f"Approximate percentage difference (using logs): {log_diff:.2f}%")

Exact percentage difference: 9.09%
Approximate percentage difference (using logs): 8.70%
